fml/task_runner_merger.cc

```diff
@@ -0,0 +1,73 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#define FML_USED_ON_EMBEDDER
+
+#include "flutter/fml/task_runner_merger.h"
+#include "flutter/fml/message_loop_impl.h"
+
+namespace fml {
+
+const int TaskRunnerMerger::kLeaseNotSet = -1;
+
+TaskRunnerMerger::TaskRunnerMerger(fml::TaskQueueId platform_queue_id,
+                                   fml::TaskQueueId gpu_queue_id)
+    : platform_queue_id_(platform_queue_id),
+      gpu_queue_id_(gpu_queue_id),
+      task_queues_(fml::MessageLoopTaskQueues::GetInstance()),
+      lease_term_(kLeaseNotSet) {
+  is_merged_ = task_queues_->Owns(platform_queue_id_, gpu_queue_id_);
+}
+
+void TaskRunnerMerger::MergeWithLease(size_t lease_term) {
+  FML_DCHECK(lease_term > 0) << "lease_term should be positive.";
+  if (!is_merged_) {
+    is_merged_ = task_queues_->Merge(platform_queue_id_, gpu_queue_id_);
+    lease_term_ = lease_term;
+  }
+}
+
+bool TaskRunnerMerger::OnWrongThread() {
+  const auto current_queue_id = MessageLoop::GetCurrentTaskQueueId();
+  if (is_merged_) {
+    return current_queue_id != platform_queue_id_;
+  } else {
+    return current_queue_id != gpu_queue_id_;
+  }
+}
+
+void TaskRunnerMerger::ExtendLease(size_t lease_term) {
```

Comment:
nit: "extend" made me wonder whether you're adding `lease_term` to the current remaining lease, maybe call it `SetLease`? (do we really care about making sure this makes the lease bigger?)

Reply:
The guarantee of this method to the caller is that the threads will stay merged for at least the `lease_term` they call with. If another caller tries to reduce the current lease term, we should still have the bigger lease term. In the existing implementation we do not ever call this with a smaller lease term than before, but I think of it as extending as we would never want to lower the lease term.